### PR TITLE
Do not use reflection for enum in non-compact mode.

### PR DESCRIPTION
**Pull request by congt**
Making wire protos j2objc-friendly. 

CC @swankjesse

### CODE REVIEW

File: wire-runtime/src/main/java/com/squareup/wire/EnumAdapter.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * An abstract {@link ProtoAdapter} that encodes values of an enum into integers.
+ */
+public abstract class EnumAdapter<E extends WireEnum> extends ProtoAdapter<E> {
+  protected EnumAdapter(Class<E> type) {
+    super(FieldEncoding.VARINT, type);
+  }
+
+  @Override public int encodedSize(E value) {
```
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/EnumAdapter.java]: `final`

File: wire-runtime/src/main/java/com/squareup/wire/EnumAdapter.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    return ProtoWriter.varint32Size(value.getValue());
+  }
+
+  @Override public void encode(ProtoWriter writer, E value) throws IOException {
```
Comment (JakeWharton) [inline wire-runtime/src/main/java/com/squareup/wire/EnumAdapter.java]: `final`

File: wire-runtime/src/test/proto-java/com/squareup/wire/protos/person/Person.java [hunk 3]
```diff
@@ -219,6 +220,22 @@ public static PhoneType fromValue(int value) {
     public int getValue() {
       return value;
     }
+
+    private static final class ProtoAdapter_PhoneType extends EnumAdapter<PhoneType> {
```
Comment (danrice-square) [inline wire-runtime/src/test/proto-java/com/squareup/wire/protos/person/Person.java]: Do we use this naming convention elsewhere? If not, I would prefer something like PhoneTypeEnumAdapter that's more human-friendly.

Comment (congt) [inline wire-runtime/src/test/proto-java/com/squareup/wire/protos/person/Person.java]: Yes, it follows the naming convention of message adapter.

File: wire-runtime/src/test/proto-java/com/google/protobuf/FieldDescriptorProto.java [hunk 4]
```diff
@@ -408,6 +409,22 @@ public static Type fromValue(int value) {
...
       return value;
     }
+
+    private static final class ProtoAdapter_Type extends EnumAdapter<Type> {
+      ProtoAdapter_Type() {
+        super(Type.class);
+      }
+
+      @Override
+      public Type decode(ProtoReader reader) throws IOException {
```
Comment (JakeWharton) [inline wire-runtime/src/test/proto-java/com/google/protobuf/FieldDescriptorProto.java]: How about pushing this implementation down into `EnumAdapter` and just adding `public abstract T fromValue(int)` which gets overridden. Then this implementation just becomes:
```java
@Override public Type fromValue(int value) {
  return Type.fromValue(value);
}
```

File: wire-runtime/src/test/proto-java/com/squareup/wire/protos/person/Person.java (hunk 3 above)
Comment (JakeWharton) [inline wire-runtime/src/test/proto-java/com/squareup/wire/protos/person/Person.java]: These classes are private and explicitly not for humans.

File: wire-runtime/src/test/proto-java/com/google/protobuf/FieldDescriptorProto.java (hunk 4 above)
Comment (congt) [inline wire-runtime/src/test/proto-java/com/google/protobuf/FieldDescriptorProto.java]: Good point!

File: wire-runtime/src/test/proto-java/com/squareup/wire/protos/person/Person.java (hunk 3 above)
Comment (danrice-square) [inline wire-runtime/src/test/proto-java/com/squareup/wire/protos/person/Person.java]: That's fine, I just hate having to look at adapter code, but I guess that is the price we paid to get rid of reflection.

File: wire-runtime/src/test/proto-java/com/google/protobuf/FieldDescriptorProto.java (hunk 4 above)
Comment (congt) [inline wire-runtime/src/test/proto-java/com/google/protobuf/FieldDescriptorProto.java]: Done

**Comment by JakeWharton** [comment]
LGTM!

**Pull request merged**